lib/pbench/server/api/resources/query_apis/__init__.py

```diff
@@ -350,3 +352,185 @@ def _get(self, json_data: JSON, _) -> Response:
+
+        Args:
+            json_data: Type-normalized client JSON input
+            _: Original incoming Request object (not used)
+
+        Returns:
+            Response to return to client
+        """
+        klasname = self.__class__.__name__
+
+        try:
+            dataset = Dataset.attach(
+                controller=json_data["controller"], name=json_data["name"]
+            )
+        except DatasetNotFound as e:
+            abort(HTTPStatus.NOT_FOUND, message=str(e))
+
+        owner = User.query(id=dataset.owner_id)
+        if not owner:
+            self.logger.error(
+                "Dataset owner ID {} cannot be found in Users", dataset.owner_id
+            )
+            abort(HTTPStatus.INTERNAL_SERVER_ERROR, message="Dataset owner not found")
+
+        # For bulk Elasticsearch operations, we check authorization against the
+        # ownership of a designated dataset rather than having an explicit
+        # "user" JSON parameter. This will raise UnauthorizedAccess on failure.
+        try:
+            self._check_authorization(owner.username, json_data["access"])
+        except UnauthorizedAccess as e:
+            abort(HTTPStatus.FORBIDDEN, message=str(e))
+
+        # Build an Elasticsearch instance to manage the bulk update
+        elastic = elasticsearch.Elasticsearch([self.es_url])
+
+        try:
+            # Pass the bulk command generator to the helper
+            results = elasticsearch.helpers.streaming_bulk(
+                elastic,
+                self.generate_documents(json_data, dataset),
+                raise_on_exception=False,
+                raise_on_error=False,
+            )
+        except Exception as e:
+            self.logger.exception(
+                "{}: exception {} occurred during the Elasticsearch request",
+                klasname,
+                type(e).__name__,
+            )
+            abort(HTTPStatus.INTERNAL_SERVER_ERROR, message="INTERNAL ERROR")
+
+        report = {}
+        count = 0
+        error_count = 0
+        for ok, response in results:
+            count += 1
+            u = response["update"]
+            status = "ok"
+            if "error" in u:
+                e = u["error"]
+                status = e["type"]
+                self.logger.debug(
+                    "{} ({}: {}) for id {} in index {}",
+                    u["status"],
+                    status,
+                    e["reason"],
+                    u["_id"],
+                    u["_index"],
+                )
+                error_count += 1
+            cnt = report.get(status, 0)
+            report[status] = cnt + 1
+
+        self.logger.info(
+            "Update access for dataset {}: {} successful document updates and {} failures",
+            dataset,
+            count - error_count,
+            error_count,
+        )
+
+        # Let the subclass complete the operation
+        self.complete(dataset, json_data, error_count)
```

Comment:
I feel there are lot of different nomenclature that we use in our query APIs like preprocess, assemble, postprocess, now complete and generate documents. In theory I feel their end-goal is similar like here `complete` is somewhat similar to `postprocess` isn't it? Do you think we need to streamline these naming conventions

Reply:
No, `complete` isn't like `postprocess` because this is focused on setting a final `Dataset` SQL state (in this case changing the `access` column) whereas "postprocess" was specifically targeted to handling the results of an Elasticsearch query ... which, in this case, is handled in common superclass code.

Really, I chose each set of names to correspond to the "missing" operations in the superclass that needed to be implemented by the subclass. For general Elasticsearch queries we needed to "assemble" the Elasticsearch query payload and "postprocess" the Elasticsearch response. (And then I complicated it by adding "preprocess", which didn't really fit but I needed a name for setup that ran before "asssemble"...)

I did initially retrofit the new publish on top of `ElasticBase` and just made `assemble` a generator and consumed the `streaming_bulk` response generator inside `postprocess`: but that results in a somewhat different signature and substantially different semantic, which is potentially confusing. Furthermore, I don't have any use for most of `ElasticBase`. It was easier to split the class tree, and reusing the same names (with different signature and somewhat subtly different semantics) also risks confusion and seemed to have no real upside.